Collect column statistics on write

arhimondr opened the pull request.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2408,9 +2408,13 @@ else if (target instanceof DeleteHandle) {
 
     private static Function<Page, Page> enforceLayoutProcessor(List<Symbol> expectedLayout, Map<Symbol, Integer> inputLayout)
     {
-        int[] channels = expectedLayout.stream()
```
you could use https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer-
to validate values withing stream. Should be cleaner

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2408,9 +2408,13 @@ else if (target instanceof DeleteHandle) {
+        for (int i = 0; i < expectedLayout.size(); i++) {
+            Symbol symbol = expectedLayout.get(i);
+            Integer channel = inputLayout.get(symbol);
+            checkArgument(channel != null, "channel not found for symbol: %s", symbol);
```
When this can happen?

sopel39 commented:
no analyze syntax yet, right? Will it be part of this PR?

arhimondr commented:
> no analyze syntax yet, right? Will it be part of this PR?

This PR is about collecting column statistics on `WRITE`. The `ANALYZE` statement will be introduced in a separate PR.

rschlussel2 commented on presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java:
```diff
@@ -141,11 +183,17 @@ public void addInput(Page page)
         state = State.FINISHED;
 
+        List<ComputedStatistics> statistics = getComputedStatistics();
+        System.out.println(statistics);
```
leftover debugging line?  But it doesn't look like `statistics` gets used anywhere else.  I'm confused about how this works.

arhimondr replied:
It is replaced in a very next commit. I'm going to remove it actually.

rschlussel2 commented on presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java:
```diff
@@ -131,15 +141,20 @@ public OperatorFactory duplicate()
         this.pageSinkMemoryContext = operatorContext.newLocalSystemMemoryContext();
         this.pageSink = requireNonNull(pageSink, "pageSink is null");
-        this.inputChannels = requireNonNull(inputChannels, "inputChannels is null");
+        this.columnChannels = requireNonNull(columnChannels, "columnChannels is null");
```
Renamed because it's input + column stats channels?

arhimondr replied:
Yup

rschlussel2 commented on presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatisticType.java:
```diff
@@ -0,0 +1,25 @@
+    NUMBER_OF_NON_NULL_VALUES,
+    MAX_VALUE_SIZE_IN_BYTES,
+    AVERAGE_VALUE_SIZE_IN_BYTES,
+    NUMBER_OF_TRUE,
```
NUMBER_OF_TRUES

electrum commented:
Typo "statistcs" in commit message

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/BridgingHiveMetastore.java:
```diff
@@ -85,33 +81,13 @@ public BridgingHiveMetastore(HiveMetastore delegate)
-                .map(Column::getName)
-                .collect(toImmutableSet());
-        return groupStatisticsByColumn(delegate.getTableColumnStatistics(databaseName, tableName, dataColumns));
+        return delegate.getTableColumnStatistics(databaseName, tableName);
```
We override the HiveMetastore` but not the `ExtendedHiveMetastore` in the internal connector. So i had to move this abstraction a level below, to do not make a mess.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -199,6 +201,8 @@ public void flushCache()
         partitionCache.invalidateAll();
         partitionFilterCache.invalidateAll();
         userTablePrivileges.invalidateAll();
+        tableStatisticsCache.invalidateAll();
```
Yes. `flush` method is a service method exposed via JNDI that flushes all the metastore caches. For some reson statistic cashes flush wasn't there. I did't think this change worth a separate commit.

arhimondr replied:
https://github.com/prestodb/presto/pull/10970

kokosing replied:
Thanks!

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -273,22 +276,39 @@ else if (table.getTableType().equals(EXTERNAL_TABLE.name())) {
+    }
+
+    @Override
+    public synchronized PartitionStatistics getTableStatistics(String databaseName, String tableName)
```
All the public methods in this class are synchronized. I guess the intention was to guarantee consistency for a single coordinator for a single file metastore case.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -1817,21 +1855,23 @@ public String toString()
                 Optional<List<String>> fileNames,
                 boolean ignoreExisting,
-                Optional<HiveBasicStatistics> statisticsUpdate)
+                PartitionStatistics statistics, PartitionStatistics statisticsUpdate)
```
`statistic` - are the updated statistics to be returned for a subsequent `getStatistics` call withing the same transaction

`statisticsUpdate` - is basically a statistics delta. It is needed when we update statistics on commit, to be able to atomically increment them.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2093,10 +2094,8 @@ public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPl
-                    .put(node.getOutputSymbols().get(0), 0)
-                    .put(node.getOutputSymbols().get(1), 1)
-                    .build();
+            Map<Symbol, Integer> layout = IntStream.range(0, node.getOutputSymbols().size()).boxed()
```
Yup. Table writer node now outputs the pre-aggregated statistics

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -84,6 +84,8 @@
 import static com.facebook.presto.hive.metastore.PrincipalType.USER;
```
I just realized that what i'm doing here doesn't make much sense. I should handle a proper schema migration as we do in PageSource and PageSink. Please ignore this commit for now. I'm going to implement proper schema migration as a separate PR.

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -325,23 +358,214 @@ private RelationPlan createTableWriterPlan(
+                return createAggregation(QualifiedName.of("count_if"), input.toSymbolReference(), BOOLEAN, BIGINT);
+            }
+            default:
+                throw new IllegalArgumentException("Unsupported statisticType: " + statisticType);
```
Nit: "statistic type"

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -325,23 +358,214 @@ private RelationPlan createTableWriterPlan(
+        }
+    }
+
+    private static boolean isApproxDistinctAvailable(Type inputType)
```
This can be removed now

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -325,23 +358,214 @@ private RelationPlan createTableWriterPlan(
+                    false,
+                    metadata.getTypeManager().isTypeOnlyCoercion(inputType, resolvedType));
+        }
+        else {
```
Could move this to the initializer

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -325,23 +358,214 @@ private RelationPlan createTableWriterPlan(
+        return new RelationPlan(commitNode, analysis.getRootScope(), commitOutputs);
+    }
+
+    private CreateStatisticAggregationsResult createStatisticsAggregation(TableStatisticsMetadata statisticsMetadata, Map<String, Symbol> columnToSymbolMap)
```
`LogicalPlanner` is already fairly long. See if you can move these methods and result callers into a separate helper class (in a different file)

arhimondr replied:
Extracted to `StatisticsAggregationsPlanner`

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java:
```diff
@@ -128,6 +136,79 @@ public TopNNode map(TopNNode node, PlanNode source, PlanNodeId newNodeId)
+        return map(node, source, node.getId());
+    }
+
+    public TableWriterNode map(TableWriterNode node, PlanNode source, PlanNodeId newNodeId)
```
I tried to follow the planning bits but I don't understand this at all. Ask @martint, @kokosing or @sopel39 to review this commit.

electrum commented on presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java:
```diff
@@ -210,18 +227,57 @@ public void addInput(Page page)
+            int positionCount = aggregationOutput.getPositionCount();
+            Block[] outputBlocks = new Block[types.size()];
+            for (int channel = 0; channel < types.size(); channel++) {
+                if (channel == ROW_COUNT_CHANNEL || channel == FRAGMENT_CHANNEL) {
```
These constants seem to be a false abstraction. We can't change them to be, say `3` and `4`, because that breaks the `channel - 2` below. So maybe it's better to just say `if (channel < 2)` here. Or add a new constant `WRITER_CHANNELS = 2`  and use that.

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2082,6 +2087,46 @@ public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPl
             PhysicalOperation source = node.getSource().accept(this, context);
 
+            ImmutableMap.Builder<Symbol, Integer> outputMapping = ImmutableMap.builder();
+            // partialrows
```
Use the constants for the channels, then you can remove the comments

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2082,6 +2087,46 @@ public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPl
+
+            OperatorFactory statisticsAggregation = node.getStatisticsAggregation().map(aggregation -> {
+                List<Symbol> groupingKeys = aggregation.getGroupingSets()
+                        .stream()
```
Nit: `stream()` on previous line

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2082,6 +2087,46 @@ public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPl
+                            source,
+                            context);
+                }
+                else {
```
Nit: else is redundant here

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2092,20 +2137,65 @@ public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPl
+
+            OperatorFactory statisticsAggregation = node.getStatisticsAggregation().map(aggregation -> {
+                List<Symbol> groupingKeys = aggregation.getGroupingSets()
+                        .stream()
```
Nit: stream on previous line

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2092,20 +2137,65 @@ public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPl
+                            source,
+                            context);
+                }
+                else {
```
Nit: else is redundant

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2092,20 +2137,65 @@ public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPl
+            Map<Symbol, Integer> aggregationOutput = outputMapping.build();
+            StatisticAggregationsDescriptor<Integer> descriptor = node.getStatisticsAggregationDescriptor()
+                    .map(desc -> desc.map(aggregationOutput::get))
+                    .orElse(StatisticAggregationsDescriptor.<Integer>builder().build());
```
Maybe add a `empty()` method to `StatisticAggregationsDescriptor`

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -2296,20 +2386,30 @@ private AccumulatorFactory buildAccumulatorFactory(
             ImmutableMap.Builder<Symbol, Integer> outputMappings = ImmutableMap.builder();
-            List<AccumulatorFactory> accumulatorFactories = new ArrayList<>();
-            for (Map.Entry<Symbol, Aggregation> entry : node.getAggregations().entrySet()) {
+            AggregationOperatorFactory operatorFactory = createAggregationOperatorFactory(node.getId(), node.getAggregations(), node.getStep(), 0, outputMappings, source, context);
```
Wrap arguments to be consistent with above

electrum commented on presto-main/src/test/java/com/facebook/presto/operator/TestTableWriterOperator.java:
```diff
@@ -59,6 +69,8 @@
 public class TestTableWriterOperator
 {
     private static final ConnectorId CONNECTOR_ID = new ConnectorId("testConnectorId");
+    private static final InternalAggregationFunction LONG_MAX = MetadataManager.createTestMetadataManager().getFunctionRegistry().getAggregateFunctionImplementation(
```
Nit: static import

electrum commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:
```diff
@@ -293,7 +294,7 @@ default ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, Co
      * Finish a table creation with data after the data is written.
      */
-    default Optional<ConnectorOutputMetadata> finishCreateTable(ConnectorSession session, ConnectorOutputTableHandle tableHandle, Collection<Slice> fragments)
+    default Optional<ConnectorOutputMetadata> finishCreateTable(ConnectorSession session, ConnectorOutputTableHandle tableHandle, Collection<Slice> fragments, List<ComputedStatistics> computedStatistics)
```
Make these `Collection` since the ordering doesn't matter

Use *Refactor -> Change Signature* to do this easily

electrum commented on presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java:
```diff
@@ -320,7 +321,7 @@ default ConnectorInsertTableHandle beginInsert(ConnectorSession session, Connect
      * Finish insert query
      */
-    default Optional<ConnectorOutputMetadata> finishInsert(ConnectorSession session, ConnectorInsertTableHandle insertHandle, Collection<Slice> fragments)
+    default Optional<ConnectorOutputMetadata> finishInsert(ConnectorSession session, ConnectorInsertTableHandle insertHandle, Collection<Slice> fragments, List<ComputedStatistics> computedStatistics)
```
Same here

electrum commented on presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java:
```diff
@@ -249,6 +248,6 @@ public void close()
     public interface TableFinisher
     {
-        Optional<ConnectorOutputMetadata> finishTable(Collection<Slice> fragments);
+        Optional<ConnectorOutputMetadata> finishTable(Collection<Slice> fragments, List<ComputedStatistics> computedStatistics);
```
Make this `Collection` here, too

electrum commented on presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java:
```diff
@@ -190,7 +191,7 @@
      * Finish a table creation with data after the data is written.
      */
-    Optional<ConnectorOutputMetadata> finishCreateTable(Session session, OutputTableHandle tableHandle, Collection<Slice> fragments);
+    Optional<ConnectorOutputMetadata> finishCreateTable(Session session, OutputTableHandle tableHandle, Collection<Slice> fragments, List<ComputedStatistics> computedStatistics);
```
Make `Collection` here, too

electrum commented on presto-hive/src/test/java/com/facebook/presto/hive/util/TestStatistics.java:
```diff
@@ -43,4 +51,104 @@ public void testReduce()
+    @Test
+    public void testMergeHiveColumnStatistics()
+    {
+        assertMergeHiveColumnStatistics(
```
I think this method actually hurts readability, since it's not clear what is happening or which is the expected output. Try inlining it.

arhimondr replied:
The idea of this method is to check that the merge is commutative. No idea how did i end up with a single assert within.

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -228,7 +238,8 @@ public HiveMetadata(
             HiveStatisticsProvider hiveStatisticsProvider,
-            int maxPartitions)
+            int maxPartitions,
+            CollectibleStatisticsProvider collectibleStatisticsProvider)
```
Let's move this above, to be after `HiveStatisticsProvider`

arhimondr replied:
`CollectibleStatisticsProvider` has been removed in favor of `ExtendedHiveMetastore#getSupportedColumnStatistics`.

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -212,6 +221,7 @@
     private final String prestoVersion;
     private final HiveStatisticsProvider hiveStatisticsProvider;
     private final int maxPartitions;
+    private final CollectibleStatisticsProvider collectibleStatisticsProvider;
```
Let's move this above, to be after `HiveStatisticsProvider`

arhimondr replied:
`CollectibleStatisticsProvider` has been removed in favor of `ExtendedHiveMetastore#getSupportedColumnStatistics`.

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -911,14 +923,18 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
         }
 
+        Map<String, Type> columnTypes = handle.getInputColumns()
+                .stream()
```
Nit: move stream to previous line

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -911,14 +923,18 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
+                    .map(PartitionUpdate::getStatistics)
+                    .reduce((first, second) -> reduce(first, second, ADD))
+                    .orElse(createZeroStatistics());
+            tableStatistic = createPartitionStatistics(session, basicStatistics, ImmutableList.of(), columnTypes, partitionComputedStatistics);
```
`tableStatistics`?

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1118,6 +1135,15 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
         }
 
+        List<String> partitionedBy = table.get().getPartitionColumns()
+                .stream()
```
Nit: stream on previous line

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1189,6 +1216,26 @@ private Partition buildPartitionObject(ConnectorSession session, Table table, Pa
+        Map<ColumnStatisticMetadata, Block> computedColumnStatistics = Optional.ofNullable(partitionComputedStatistics.get(partitionValues))
+                .map(ComputedStatistics::getColumnStatistics)
+                .orElse(ImmutableMap.of());
+        verify(basicStatistics.getRowCount().isPresent(), "rowCount is not present");
```
Combine the extract and check
```java
long rowCount = basicStatistics.getRowCount().orElseThrow(() -> new IllegalArgumentException("rowCount not present"));

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java:
```diff
@@ -601,4 +601,18 @@ public static Decimal toMetastoreDecimal(BigDecimal decimal)
+            }
+            return distinctValuesCount;
+        }
+        return OptionalLong.empty();
```
This seems wrong. If `distinctValuesCount` is present but `nullsCount` is missing, we return empty. If that is intentional, document why.

I would expect the algorithm to be
```java
if (distinctValuesCount.isPresent() && nullsCount.isPresent() && (nullsCount.getAsLong() > 0)) {
    return OptionalLong.of(distinctValuesCount.getAsLong() + 1);
}
return distinctValuesCount;
```

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -59,22 +94,72 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
 
-    public static boolean isMinMaxSupportedForType(Type type)
+    public static Map<String, HiveColumnStatistics> merge(
+            Map<String, HiveColumnStatistics> first, OptionalLong firstRowCount, Map<String, HiveColumnStatistics> second, OptionalLong secondRowCount)
```
Nit: wrap all arguments or none

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -59,22 +94,72 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
-                || type.equals(DATE)
-                || type.equals(TIMESTAMP)
-                || (type instanceof DecimalType);
+        // skip the columns for which statistics from either of sides are missing
```
How about
```java
// only keep columns that have statistics for both sides

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -59,22 +94,72 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
+        // skip the columns for which statistics from either of sides are missing
+        Set<String> columns = intersection(first.keySet(), second.keySet());
+        return columns.stream()
+                .collect(toImmutableMap(column -> column, column -> merge(first.get(column), firstRowCount, second.get(column), secondRowCount)));
```
Wrap the arguments to `toImmutableMap` since this is long

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -59,22 +94,72 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
+        if (type.equals(BOOLEAN)) {
+            return ImmutableSet.of(NUMBER_OF_NON_NULL_VALUES, NUMBER_OF_TRUE);
+        }
+        else if (type.equals(BIGINT) || type.equals(INTEGER) || type.equals(SMALLINT) || type.equals(TINYINT)) {
```
Nit: else are redundant

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -59,22 +94,72 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
+        else if (type.equals(BIGINT) || type.equals(INTEGER) || type.equals(SMALLINT) || type.equals(TINYINT)) {
+            return ImmutableSet.of(NUMBER_OF_NON_NULL_VALUES, MIN, MAX, NUMBER_OF_DISTINCT_VALUES);
+        }
+        else if (type.equals(DOUBLE) || type.equals(REAL)) {
```
Let's combine the branches all of these numeric-ish types (including data/timestamp), so it's clear we are collecting the same stats

arhimondr replied:
Changed to `if (isNumericType(type) || type.equals(DATE) || type.equals(TIMESTAMP)) {`

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+    private static List<String> getPartitionValues(ComputedStatistics statistics, List<String> partitionColumns, List<Type> partitionColumnTypes)
+    {
+        checkArgument(statistics.getGroupingColumns().equals(partitionColumns),
+                "Unexpected groping. Partition columns: %s. Grouping columns: %s", partitionColumns, statistics.getGroupingColumns());
```
Typo "groping"

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+    public static Map<String, HiveColumnStatistics> fromComputedStatistics(
+            ConnectorSession session, DateTimeZone timeZone, Map<ColumnStatisticMetadata, Block> computedStatistics, Map<String, Type> columnTypes, long rowCount)
+    {
+        return groupByColumn(computedStatistics)
```
Nit: put `entrySet().stream()` on this line

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+    }
+
+    public static Map<String, HiveColumnStatistics> fromComputedStatistics(
+            ConnectorSession session, DateTimeZone timeZone, Map<ColumnStatisticMetadata, Block> computedStatistics, Map<String, Type> columnTypes, long rowCount)
```
Wrap all parameters or none

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+    private static Map<String, Map<ColumnStatisticType, Block>> groupByColumn(Map<ColumnStatisticMetadata, Block> computedStatistics)
+    {
+        Map<String, Map<ColumnStatisticType, Block>> result = new HashMap<>();
+        computedStatistics.forEach((metadata, block) -> {
```
You might be able to use `Collectors.groupingBy` here

arhimondr replied:
I tried, but it really looks like an obfuscation then.

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+
+        // NDV
+        if (computedStatistics.containsKey(NUMBER_OF_DISTINCT_VALUES)) {
+            result.setDistinctValuesCount(BigintType.BIGINT.getLong(computedStatistics.get(NUMBER_OF_DISTINCT_VALUES), 0));
```
Static import `BIGINT`

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+        if (type.equals(BIGINT) || type.equals(INTEGER) || type.equals(SMALLINT) || type.equals(TINYINT)) {
+            return ((Number) type.getObjectValue(session, block, 0)).longValue();
+        }
+        else if (type.equals(DOUBLE) || type.equals(REAL)) {
```
Nit: else is redundant

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+            return LocalDate.ofEpochDay(((SqlDate) type.getObjectValue(session, block, 0)).getDays());
+        }
+        else if (type.equals(TIMESTAMP)) {
+            long valueUtc = block.getLong(0, 0);
```
I'd inline this since it's short and the variable doesn't add much value

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -115,6 +200,120 @@ else if (prestoType instanceof DecimalType) {
+        }
+        else if (type.equals(TIMESTAMP)) {
+            long valueUtc = block.getLong(0, 0);
+            return timeZone.convertUTCToLocal(valueUtc) / 1000;
```
Maybe do this as
```java
return MILLISECONDS.toSeconds(...)

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -123,14 +322,78 @@ private static OptionalLong reduce(OptionalLong first, OptionalLong second, Redu
+        if (firstOptional.isPresent() && secondOptional.isPresent()) {
+            Comparable<?> first = firstOptional.get();
+            Comparable<?> second = secondOptional.get();
+            checkArgument(first.getClass().equals(second.getClass()), "cannot compare comparable of different types: %s != %s", first.getClass(), second.getClass());
```
Nit: use `first.getClass().getName()` to avoid the `"class "` prefix from the `toString()` method

arhimondr replied:
The method was removed

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -123,14 +322,78 @@ private static OptionalLong reduce(OptionalLong first, OptionalLong second, Redu
+
+    private static <T extends Comparable<T>> T max(T first, T second)
+    {
+        return first.compareTo(second) > 0 ? first : second;
```
Might be better to use `>=` so that we return `first` on tie (even though they should be equal ... returning first seems more natural)

electrum commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -123,14 +322,78 @@ private static OptionalLong reduce(OptionalLong first, OptionalLong second, Redu
         SUBTRACT,
+        SELECT_MIN,
+        SELECT_MAX,
+        AVG
```
Nit: add trailing comma

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -258,55 +262,51 @@ public void flushCache()
                 partition.getHiveTableName().getTableName(),
                 ImmutableSet.of(partition.getPartitionName()));
-        return columnStatistics.getOrDefault(partition.getPartitionName(), ImmutableMap.of());
+        verify(partitionStatistics.containsKey(partition.getPartitionName()), "the result doesn't contain statistics for partition %s", partition.getPartitionName());
```
This is something that is never supposed to happen, as long as there is no bug in the Metastore code. I thought that we use `HIVE_METASTORE_ERROR` for the runtime metastore errores.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -473,6 +473,7 @@ protected void invalidateTable(String databaseName, String tableName)
                 .filter(userTableKey -> userTableKey.matches(databaseName, tableName))
                 .forEach(userTablePrivileges::invalidate);
         invalidatePartitionCache(databaseName, tableName);
+        tableStatisticsCache.invalidate(new HiveTableName(databaseName, tableName));
```
Yeah, i just didn't want to introduce yet another commit. There already lot of them. Let me know if you still want me to extract that.

arhimondr replied:
https://github.com/prestodb/presto/pull/10970

kokosing replied:
Thanks

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/HiveColumnStatistics.java:
```diff
@@ -49,43 +58,172 @@ public HiveColumnStatistics(
+        return toStringHelper(this)
+                .add("lowValue", lowValue)
+                .add("highValue", highValue)
+                .add("maxColumnLength", maxColumnLength)
```
I think this `toString` method is not going to be used for other than debugging purposes. I find it useful when i see in a debugger what is actually there as a result of  `toString`. Printing a field as null is misleading. Because it may be null than in two different cases. When the optional is empty, and when the field itself is null (what surprisingly easy to achieve for this very class, since there are no null checks, but i'm gonna fix that, that's not the point).

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/HiveColumnStatistics.java:
```diff
@@ -29,15 +37,16 @@
+            @JsonProperty("nullsCount") OptionalLong nullsCount,
+            @JsonProperty("distinctValuesCount") OptionalLong distinctValuesCount)
     {
         this.lowValue = lowValue;
```
Going to add the null checks as part of this commit

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -273,22 +276,39 @@ else if (table.getTableType().equals(EXTERNAL_TABLE.name())) {
+                    .orElseThrow(() -> new PartitionNotFoundException(new SchemaTableName(databaseName, tableName), ImmutableList.copyOf(partitionValues)));
+            Map<String, HiveColumnStatistics> columnStatistics = partitionMetadata.getColumnStatistics().entrySet()
+                    .stream()
+                    .collect(toImmutableMap(Entry::getKey, Entry::getValue));
```
Looks like a rebase artifact =)

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/PartitionMetadata.java:
```diff
@@ -83,6 +93,7 @@ public PartitionMetadata(Table table, Partition partition)
 
         bucketProperty = partition.getStorage().getBucketProperty();
         serdeParameters = partition.getStorage().getSerdeParameters();
+        this.columnStatistics = ImmutableMap.copyOf(requireNonNull(columnStatistics, "columnStatistics is null"));
```
I actually like to have a clear message what field exactly was null. As the code changes the line number in the stacktrace might be inaccurate.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/TableMetadata.java:
```diff
@@ -110,6 +121,7 @@ public TableMetadata(Table table)
 
         viewOriginalText = table.getViewOriginalText();
         viewExpandedText = table.getViewExpandedText();
+        this.columnStatistics = ImmutableMap.copyOf(requireNonNull(columnStatistics, "columnStatistics is null"));
```
Same answer. If you feel strongly about that i don't mind changing though.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -16,54 +16,43 @@
-        updatedParameters.putAll(toStatisticParameters(updatedStatistics));
-        return unmodifiableMap(updatedParameters);
+        HiveBasicStatistics originalStatistics = getHiveBasicStatistics(parameters);
+        HiveBasicStatistics updatedStatistics = Statistics.reduce(originalStatistics, statistics, operator);
```
No, it looks like a refactoring leftover

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/InMemoryHiveMetastore.java:
```diff
@@ -441,23 +443,25 @@ private static boolean partitionMatches(Partition partition, String databaseName
+    @Override
+    public boolean supportsColumnStatistics()
+    {
+        return false;
```
I don't know why is it still `false` in this commit. Probably a squashing-separation artifact.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/InMemoryHiveMetastore.java:
```diff
@@ -455,9 +453,7 @@ public synchronized PartitionStatistics getTableStatistics(String databaseName,
         SchemaTableName schemaTableName = new SchemaTableName(databaseName, tableName);
         PartitionStatistics statistics = columnStatistics.get(schemaTableName);
         if (statistics == null) {
-            // TODO: remove this after basic statistics are migrated to updateTableStatistics
```
I did in a previous commit. It was `false` because of some "rebase" kind of mistake.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -325,23 +358,214 @@ private RelationPlan createTableWriterPlan(
+        }
+
+        FunctionRegistry functionRegistry = metadata.getFunctionRegistry();
+        if (statisticsMetadata.getTableStatistics().size() > 0) {
```
Based on the online discussion:

- Although we don't need to compute `ROW_COUNT` when writing a table, we may want to recompute the `ROW_COUNT` with `ANALYZE`. And the SPI objects were supposed to be the same.
- Further we may want to add `UNCOMPRESSED_DATA_SIZE` statistic.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -325,23 +358,214 @@ private RelationPlan createTableWriterPlan(
+            }
+            case NUMBER_OF_DISTINCT_VALUES: {
+                checkArgument(inputType.isComparable(), "Input type is not comparable: %s", inputType);
+                // TODO: IMPLEMENT GENERIC APPROX DISTINCT
```
It is still in progress: https://github.com/prestodb/presto/pull/10899 

@jessesleeping, do you think you can merge it soon?

arhimondr commented on presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatisticMetadata.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    @JsonProperty
+    public String getColumnName()
```
Both `getInsertLayout` and `getNewTableLayout` use the same class, the `ConnectorNewTableLayout`, as a return type. I'm not sure if it is worth complications having two separate classes.  @martint

arhimondr commented on presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatisticType.java:
```diff
@@ -0,0 +1,25 @@
+{
+    MIN,
+    MAX,
+    NUMBER_OF_DISTINCT_VALUES,
```
I'm going to leave it as is to be consistent with other names.

kokosing commented on presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java:
```diff
@@ -30,6 +30,7 @@
 import com.facebook.presto.spi.security.GrantInfo;
```
I think I am missing tests like `TestLogicalPlanner` to see what plans are generated for CREATE and INSERT queries

arhimondr replied:
I was thinking about such a test, but finally i decided to do not implement one. Statistic aggregations don't change the plan shape but just add additional output symbols. 

Currently `BasePlanTest` uses the `LocalQueryRunner` with the TPCH connector installed. TPCH connector doesn't ask the engine to compute any statistics. So to do that i would have to change the TPCH connector to request statistics for some tables. Also i would need to have additional tables that request statistics to be pre-grouped. 

I found that test to be too tedious for the value it introduces. What this test can actually verify, is that the planner creates the correct aggregations for given column-statistics pairs. The regression caused by modifying that code will be caught by the product/integration tests that have all possible aggregation for the all possible types covered. We just may spend a little bit longer trying to figure out what has happened.

kokosing commented on presto-main/src/main/java/com/facebook/presto/operator/NoOpOperator.java:
```diff
@@ -0,0 +1,96 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class NoOpOperator
```
BlackHole could be confusing too, because it may refer to the connector name. `DevNullOperator`?

arhimondr replied:
Let's call it `DevNullOperator`

kokosing commented on presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java:
```diff
@@ -42,21 +45,30 @@
         private final int operatorId;
         private final PlanNodeId planNodeId;
         private final TableFinisher tableFinisher;
+        private final OperatorFactory statisticsAggregation;
```
s/statisticsAggregation/operatorFactory?

arhimondr replied:
I think `statisticsAggregation` better describes what it is in the context of the `TableFinishOperator`

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/HiveMetastore.java:
```diff
@@ -13,8 +13,8 @@
  */
 package com.facebook.presto.hive.metastore.thrift;
 
+import com.facebook.presto.hive.metastore.HiveColumnStatistics;
```
Our internal connector overrides the `HiveMetastore`, and does not override the `ExtendedHiveMetastore`. To do not make a mess it was decided to move this abstraction down.

CC: @electrum @haozhun

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -231,20 +233,25 @@ private static boolean isPrestoView(Table table)
         }
         catch (NoSuchObjectException e) {
-            return ImmutableSet.of();
+            throw new TableNotFoundException(new SchemaTableName(databaseName, tableName));
```
`client.getTableColumnStatistics` may throw it in case of a race

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -258,55 +262,55 @@ public void flushCache()
                 ImmutableSet.of(partition.getPartitionName()));
-        return columnStatistics.getOrDefault(partition.getPartitionName(), ImmutableMap.of());
+        if (!partitionStatistics.containsKey(partition.getPartitionName())) {
+            throw new PrestoException(HIVE_PARTITION_DROPPED_DURING_QUERY, "Statistics result does not contain entry for partition: " + partition.getPartitionName());
```
How does it change the semantic?

arhimondr replied:
Ok. I think i understand your question now. So before this commit this method used to return only the partition *column* statistics, which are allowed to be missing for a give partition. But now the contract has changed. If the statistics are missing, the underlying method should always return a non-null `PartitionStatistics` object for every partition requested. If the statistics are missing - it will have an empty map inside. The underlying metastore can return `null` only if the partition is no longer there -> the partition has been dropped during the execution.

kokosing replied:
It would be nice to separate semantic changes from refactorings, but I understand that managing such big PRs could be a pain.

arhimondr commented:
> I think we need to divide this PR into several smaller ones.

It has been already reviewed as a whole. Separating it now will make a muddle.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -231,20 +233,25 @@ private static boolean isPrestoView(Table table)
-    public Set<ColumnStatisticsObj> getTableColumnStatistics(String databaseName, String tableName, Set<String> columnNames)
+    public Map<String, HiveColumnStatistics> getTableColumnStatistics(String databaseName, String tableName)
     {
+        Table table = getTable(databaseName, tableName)
```
The other way around is to have a column list as an argument. But it makes the interface a way more harder to understand and complicates the caching layer. 

Also I'm going to do the schema migration for the statistics. Similar as we do for the data reads and writes. This is when the table schema is changed, but the partition schema remains the same. Than it is unclear what column name to pass. Whether the column names stored in partition, or the column names stored in the table.

kokosing replied:
> Yes, but after that the statistics object will be cached, and no further calls will be needed.

We ask for a table many times per query already.

What if you would accept a `Table` as parameter?

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -255,23 +262,30 @@ private static boolean isPrestoView(Table table)
-    public Map<String, Set<ColumnStatisticsObj>> getPartitionColumnStatistics(String databaseName, String tableName, Set<String> partitionNames, Set<String> columnNames)
+    public Map<String, Map<String, HiveColumnStatistics>> getPartitionColumnStatistics(String databaseName, String tableName, Set<String> partitionNames)
     {
+        Table table = getTable(databaseName, tableName)
```
The same answer + it might be replaced with a `Partition` once the schema evolution is implemented.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java:
```diff
@@ -258,55 +262,55 @@ public void flushCache()
-                else {
-                    result.put(HivePartitionName.partition(table, partitionName), ImmutableMap.of());
+                if (!partitionStatistics.containsKey(partitionName)) {
+                    throw new PrestoException(HIVE_PARTITION_DROPPED_DURING_QUERY, "Statistics result does not contain entry for partition: " + partitionName);
```
ditto

arhimondr commented:
I addressed all the comments from the refactor part (first 7 commits), and extracted them as a separate PR.

@electrum @kokosing Please have a final look:

https://github.com/prestodb/presto/pull/10972

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -216,12 +226,15 @@ private RelationPlan createTableCreationPlan(Analysis analysis, Query query)
                 .map(ColumnMetadata::getName)
                 .collect(toImmutableList());
 
+        TableStatisticsMetadata statisticsMetadata = metadata.getNewTableStatisticsMetadata(session, destination.getCatalogName(), tableMetadata);
```
I would imply that there is an old one. I would call it `newTableStatisticsMetadata`, but it seems to be very verbose.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java:
```diff
@@ -275,26 +288,25 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
         plan = new RelationPlan(projectNode, scope, projectNode.getOutputSymbols());
 
         Optional<NewTableLayout> newTableLayout = metadata.getInsertLayout(session, insert.getTarget());
+        TableStatisticsMetadata statisticsMetadata = metadata.getInsertStatisticsMetadata(session, insert.getTarget());
```
Apparently you want it to be consistent with the `newTableLayout`. But honestly i don't think that's a good name, since it is returned by the `getInsertLayout` method.

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:
```diff
@@ -1044,4 +1047,25 @@ public boolean isTableStatisticsEnabled()
+    {
+        ENABLED,
+        ENABLED_FOR_MARKED_TABLES,
+        DISABLED
```
put DISABLED first, end all options with `,`

arhimondr replied:
I removed the enum

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:
```diff
@@ -1044,4 +1047,25 @@ public boolean isTableStatisticsEnabled()
+    public enum CollectColumnStatisticsOnWriteOption
+    {
+        ENABLED,
+        ENABLED_FOR_MARKED_TABLES,
```
add a javadoc linking to `com.facebook.presto.hive.HiveTableProperties#COLLECT_COLUMN_STATISTICS_ON_WRITE_ENABLED`

arhimondr replied:
I removed the enum

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -438,6 +442,11 @@ private ConnectorTableMetadata getTableMetadata(SchemaTableName tableName)
         Optional<String> comment = Optional.ofNullable(table.get().getParameters().get(TABLE_COMMENT));
 
+        String collectColumnStatisticsOnWrite = table.get().getParameters().get(COLLECT_COLUMN_STATISTICS_ON_WRITE_ENABLED_KEY);
+        if (collectColumnStatisticsOnWrite != null && Boolean.valueOf(collectColumnStatisticsOnWrite)) {
```
I don't like `Boolean.valueOf` because it accepts `"yes"` but returns `false`

arhimondr replied:
Removed

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java:
```diff
@@ -49,6 +50,7 @@
     public static final String SORTED_BY_PROPERTY = "sorted_by";
     public static final String ORC_BLOOM_FILTER_COLUMNS = "orc_bloom_filter_columns";
     public static final String ORC_BLOOM_FILTER_FPP = "orc_bloom_filter_fpp";
+    public static final String COLLECT_COLUMN_STATISTICS_ON_WRITE_ENABLED = "collect_column_statistics_on_write_enabled";
```
This appears to be presto-specific property, maybe prefix it with `presto_`?

arhimondr replied:
Removed

arhimondr commented on presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java:
```diff
@@ -193,6 +194,16 @@
+    /**
+     * Describes statistics that must be collected for an existing table during the INSERT operation
+     */
+    TableStatisticsMetadata getInsertStatisticsMetadata(Session session, TableHandle tableHandle);
```
Actually i don't think we need this. We can resolve the `ConnectorTableMetadata` in the planner, and use the `getNewTableStatisticsMetadata`. 

I followed the approach used for the `getInsertLayout`, what seems to be extra in this situation. We are not going to have different statistics set for the INSERT and the CREATE TABLE cases.

I'm going to remove this method, and rename the `getNewTableStatisticsMetadata` to `getStatisticsMetadata`

arhimondr replied:
upd: `getStatisticsCollectionMetadata`

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/StatisticsAggregationPlanner.java:
```diff
@@ -0,0 +1,264 @@
+
+    private static boolean isApproxDistinctAvailable(Type inputType)
+    {
+        return inputType.equals(BIGINT) || inputType.equals(DOUBLE) || isVarcharType(inputType) || inputType.equals(VARBINARY);
```
Since we got the `approx_distinct` for all the types we need, i'm just going to remove this.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -0,0 +1,103 @@
+        return groupingSymbols;
+    }
+
+    public Parts split(SymbolAllocator symbolAllocator, FunctionRegistry functionRegistry)
```
`createIntermediateAggregations `

arhimondr commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregations.java:
```diff
@@ -0,0 +1,103 @@
+                            Optional.empty()));
+        }
+        return new Parts(
+                new StatisticAggregations(intermediateAggregation.build(), groupingSymbols),
```
I think the correct name is `intermediate`. In aggregation function declaration we have finalType and intermediateType.

findepi replied:
Have a look at `com.facebook.presto.sql.planner.plan.AggregationNode.Step` enum.
* PARTIAL creates intermediate results
* INTERMEDIATE reduces intermediate results into intermediate results
* FINAL reduces intermediate results into aggregation's final result. 

(Obviously, INTERMEDIATE is optional.)

Thus, what you do here, is to create_partial_ aggregation + _final_ aggregation.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java:
```diff
@@ -114,10 +135,16 @@ public boolean isFinished()
         return state == State.FINISHED;
```
I didn't know it may not finish upon finish call. It makes sense.

arhimondr replied:
Actually the state of this operator cannot be `FINISHED` if the aggregation operator is not finished. I think the additional check is extra here.

findepi replied:
then add some `verify` here. I find such `verify`-s helpful when reading the code. They narrow down the space of possible states much faster than reading the code & reasoning about it.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java:
```diff
@@ -114,10 +135,16 @@ public boolean isFinished()
     public boolean needsInput()
     {
-        return state == State.RUNNING;
+        return state == State.RUNNING && statisticsAggregation.needsInput();
```
I assume that the aggregation is never getting full. The spill is always disabled, so it is either fails, or accepts the page. But i agree, let it better be an explicit assertion.

arhimondr replied:
Added a comment explaining what is going on

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java:
```diff
@@ -156,6 +171,7 @@ public void finish()
             finishFuture = pageSink.finish();
             blocked = toListenableFuture(finishFuture);
             updateWrittenBytes();
+            statisticAggregation.finish();
```
State of the `TableWriterOperator` cannot be finished before the `statisticAggregation` is finished. Think of `statisticAggregation` as of an internal `TableWriterOperator` that is fully managed by the `TableWriterOperator`. `TableWriterOperator` cannot be finished if the `statisticAggregation` is not.

arhimondr replied:
`put outside (after) if` +

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java:
```diff
@@ -177,13 +193,13 @@ public boolean isFinished()
     {
         if (blocked != NOT_BLOCKED && blocked.isDone()) {
-            blocked = NOT_BLOCKED;
+            blocked = statisticAggregation.isBlocked();
```
Yeah, you are right.

I should've spent more time analyzing the interactions. Trying to solve the race with calling `isBlocked` + `allAsList`. It seems to be that now it should be fine. But i would ask you to verify it carefully as well. I could've made a mistake again.

findepi replied:
`allAsList` has its own implications that we need to be aware of -- every time it's called, it adds a listener to the underlying futures, leaking memory.
In `Driver`, we address this with being less "functional", caching the future (`com.facebook.presto.operator.Driver#driverBlockedFuture`). If you wish, I can think about this more.

arhimondr replied:
I'm well aware of that issue. Previously we had problems because we were creating a new `allAsList` every time the `isBlocked` method was called. I'm creating an `allAsList` future on `addPage` or `finish`, when the state is actually changing.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java:
```diff
@@ -210,18 +227,57 @@ public void addInput(Page page)
+            return new Page(positionCount, outputBlocks);
+        }
+
+        if (state != State.FINISHING) {
```
Yeah, that's a bit confusing. I moved around the variables a bit.

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/TableWriterOperator.java:
```diff
@@ -242,12 +298,14 @@ public Page getOutput()
             if (!committed) {
                 pageSink.abort();
             }
+            statisticAggregation.close();
```
The problem is that the `operator` is `AutoCloseable`, and not `Closeable`. So i would have to write the cumbersome `Closer` myself. But anyhow, you are right, we need to to this anyway.

arhimondr replied:
I implemented a `AutoCloseableCloser`

findepi replied:
> The problem is that the operator is AutoCloseable, and not Closeable.

This is easy to solve: `closer.register(operator::close)` instead of `closer.register(operator)`.

arhimondr replied:
Closeable#close throws the `IOException`, when Autocloseable#close throws a more generic `Exception`

arhimondr commented on presto-spi/src/main/java/com/facebook/presto/spi/statistics/ComputedStatistics.java:
```diff
@@ -0,0 +1,102 @@
+            Map<TableStatisticType, Block> tableStatistics,
+            Map<ColumnStatisticMetadata, Block> columnStatistics)
+    {
+        this.groupingColumns = unmodifiableList(requireNonNull(groupingColumns, "groupingColumns is null"));
```
Decided to go with a copy technique. We use it in the other places.

arhimondr commented on presto-spi/src/main/java/com/facebook/presto/spi/statistics/ComputedStatistics.java:
```diff
@@ -0,0 +1,102 @@
+            Map<ColumnStatisticMetadata, Block> columnStatistics)
+    {
+        this.groupingColumns = unmodifiableList(requireNonNull(groupingColumns, "groupingColumns is null"));
+        this.gropingValues = unmodifiableList(requireNonNull(gropingValues, "gropingValues is null"));
```
For now block is supposed to hold only a single value. But that might not be true for the advanced statistics in the future.

arhimondr commented on presto-spi/src/main/java/com/facebook/presto/spi/statistics/ComputedStatistics.java:
```diff
@@ -0,0 +1,102 @@
+    {
+        this.groupingColumns = unmodifiableList(requireNonNull(groupingColumns, "groupingColumns is null"));
+        this.gropingValues = unmodifiableList(requireNonNull(gropingValues, "gropingValues is null"));
+        this.tableStatistics = unmodifiableMap(requireNonNull(tableStatistics, "tableStatistics is null"));
```
It might not be true for the advanced statistics. We can check it now though.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:
```diff
@@ -60,6 +61,7 @@ public void configure(Binder binder)
         binder.bind(HiveConnectorId.class).toInstance(new HiveConnectorId(connectorId));
         binder.bind(TypeTranslator.class).toInstance(new HiveTypeTranslator());
         binder.bind(CoercionPolicy.class).to(HiveCoercionPolicy.class).in(Scopes.SINGLETON);
+        binder.bind(CollectibleStatisticsProvider.class).toInstance(Statistics::getSupportedStatistics);
```
I'm going to replace `ExtendedHiveMetastore#supportsColumnStatistics` with the `ExtendedHiveMetastore#getSupportedColumnStatistics`

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java:
```diff
@@ -310,4 +320,10 @@ private Estimate summarizePartitionStatistics(
         }
     }
+
+    private boolean isMinMaxSupportedForType(Type type)
```
This method was not used and has been removed.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java:
```diff
@@ -310,4 +320,10 @@ private Estimate summarizePartitionStatistics(
+
+    private boolean isMinMaxSupportedForType(Type type)
+    {
+        Set<ColumnStatisticType> statisticTypes = collectibleStatisticsProvider.get(type);
```
The method has been removed

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -59,22 +94,72 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
     }
 
-    public static boolean isMinMaxSupportedForType(Type type)
+    public static Map<String, HiveColumnStatistics> merge(
```
These methods are not needed in `Use updateStatisticsMethods to store basic stats`. The `Use updateStatisticsMethods to store basic stats` commit introduce methods required to store previously prepared column statistics. `merge` and `reduce` method are required to prepare the column statistics to be stored.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java:
```diff
@@ -119,14 +318,78 @@ private static OptionalLong reduce(OptionalLong first, OptionalLong second, Redu
+            Optional<? extends Comparable<?>> secondOptional,
+            ReduceOperator operator)
+    {
+        if (firstOptional.isPresent() && secondOptional.isPresent()) {
```
Yes. Statistics may be missing in the Metastore

arhimondr replied:
I removed this method. We no longer have to deal with untyped min and max.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -773,6 +773,14 @@ private void testCreateInvalidBucketedTable(HiveStorageFormat storageFormat)
+    public void testCreatePartitionedUnionAll()
+    {
+        assertUpdate("CREATE TABLE test_create_partitioned_union_all (a varchar, ds varchar) WITH (partitioned_by = ARRAY['ds'])");
+        assertUpdate("INSERT INTO test_create_partitioned_union_all SELECT 'a', '2013-05-17' UNION ALL SELECT 'b', '2013-05-17'", 2);
```
Integration test for the `PushTableWriteThroughUnion`

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java:
```diff
@@ -134,6 +135,8 @@
 
     private boolean tableStatisticsEnabled = true;
 
+    private CollectColumnStatisticsOnWriteOption collectColumnStatisticsOnWrite = ENABLED_FOR_MARKED_TABLES;
```
Actually i think meanwhile the plan had changed, and we no longer need this. I will replace the enum with a simple boolean.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -204,6 +206,8 @@
     private static final String ORC_BLOOM_FILTER_COLUMNS_KEY = "orc.bloom.filter.columns";
     private static final String ORC_BLOOM_FILTER_FPP_KEY = "orc.bloom.filter.fpp";
 
+    private static final String COLLECT_COLUMN_STATISTICS_ON_WRITE_ENABLED_KEY = "presto.collect.column.statistics.on.write.key";
```
Removed

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -686,6 +695,8 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
         // Table comment property
         tableMetadata.getComment().ifPresent(value -> tableProperties.put(TABLE_COMMENT, value));
 
+        tableProperties.put(COLLECT_COLUMN_STATISTICS_ON_WRITE_ENABLED_KEY, Boolean.toString(HiveTableProperties.isCollectColumnStatisticsOnWriteEnabled(tableMetadata.getProperties())));
```
Removed

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1582,10 +1596,28 @@ public TableStatisticsMetadata getInsertIntoTableStatisticsMetadata(ConnectorSes
+            case ENABLED:
+                return true;
+            case ENABLED_FOR_MARKED_TABLES:
+                return HiveTableProperties.isCollectColumnStatisticsOnWriteEnabled(tableMetadata.getProperties());
```
Removed

arhimondr commented:
Superseded by: #11054
